### PR TITLE
Add pagination to recipes

### PR DESCRIPTION
**Pull request by TomaszLaski**
added pagination, it is inside RecipesList folder becouse for now it is not suitable for reusability; add bootstrap for styles; changed component structure and added simple loader

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -12,7 +13,9 @@
     "@types/react": "^17.0.0",
     "@types/react-dom": "^17.0.0",
     "axios": "^0.21.1",
+    "bootstrap": "^5.1.0",
```
Comment (Aliath) [inline package.json:16]: Huh, once `material-ui` is picked as UI library we should not mix it with bootstrap

File: src/components/RecipesList/RecipesDownloader.tsx [hunk 2]
```diff
@@ -0,0 +1,63 @@
+import React from 'react';
+import ListItem from '@material-ui/core/ListItem';
+import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
+import ListItemText from '@material-ui/core/ListItemText';
+import { List, Grid } from '@material-ui/core';
+import Paper from '@material-ui/core/Paper';
+import { Recipe } from 'services/recipes';
+import CircularProgress from '@material-ui/core/CircularProgress';
+
+interface Props {
```
Comment (Aliath) [inline src/components/RecipesList/RecipesDownloader.tsx:10]: `Props` is not enough specific

Comment (Aliath) [reply src/components/RecipesList/RecipesDownloader.tsx:10]:
```suggestion
interface RecipesDownloaderProps {
```

File: src/components/RecipesList/RecipesDownloader.tsx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    },
+    textList: {
+      textAlign: 'center',
+      width: '100%',
+      padding: '0.3rem',
+    },
+  })
+);
+
+export const RecipesDownloader: React.FC<Props> = ({ loading, recipes }) => {
```
Comment (Aliath) [inline src/components/RecipesList/RecipesDownloader.tsx:31]: `Downloader` sounds like it download some file on user computer - rename it

File: src/components/RecipesList/RecipesList.tsx [hunk 4]
```diff
@@ -19,36 +26,89 @@ const useStyles = makeStyles(() =>
...
-            return (
-              <ListItem alignItems="flex-start" key={recipe.id}>
-                <Paper className={recipesStyles.textList}>
-                  <ListItemText primary={recipe.name} />
-                </Paper>
-              </ListItem>
-            );
-          })}
-        </List>
+        <form>
```
Comment (Aliath) [inline src/components/RecipesList/RecipesList.tsx:78]: If you will never submit this form it probably should not be a form 😄

File: src/components/RecipesList/RecipesList.tsx [hunk 5]
```diff
@@ -19,36 +26,89 @@ const useStyles = makeStyles(() =>
...
+          <label htmlFor="recipes-num">Number of recipes per page:</label>
+          <input
```
Comment (Aliath) [inline src/components/RecipesList/RecipesList.tsx:80]: What about [this](https://material-ui.com/components/pagination/)?

File: src/components/RecipesList/RecipesList.tsx [hunk 6]
```diff
@@ -19,36 +26,89 @@ const useStyles = makeStyles(() =>
...
+        </form>
+
+        <Button
+          variant="outlined"
+          color="primary"
+          onClick={() => setRecipesPerPage(allRecipes)}
+        >
+          Show all Recipes
+        </Button>
+        <RecipesDownloader loading={loading} recipes={currentRecipes} />
```
Comment (Aliath) [inline src/components/RecipesList/RecipesList.tsx:106]: Can you explain me what actually `RecipesDownloader` is used for?

File: src/components/RecipesList/RecipesPagination.tsx [hunk 7]
```diff
@@ -0,0 +1,38 @@
+import React from 'react';
+import 'bootstrap/dist/css/bootstrap.min.css';
+
+interface Props {
+  recipesPerPage: number;
+  totalRecipes: number;
+  paginate: any;
+}
+
+const RecipesPagination: React.FC<Props> = ({
```
Comment (Aliath) [inline src/components/RecipesList/RecipesPagination.tsx:10]: I don't think that custom pagination is something that we want to use:
- we have material-ui component dedicated for this
- we have to test custom component if it works well
- we should care about pages generating (for loop) when you actually do not have to 😎

File: src/components/RecipesList/RecipesPagination.tsx [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+              {number}
+            </button>
+          </li>
+        ))}
+      </ul>
+    </nav>
+  );
+};
+
+export default RecipesPagination;
```
Comment (Aliath) [inline src/components/RecipesList/RecipesPagination.tsx:38]: What is the target of `export default` over here? 😄

File: src/components/RecipesList/RecipesList.tsx [hunk 9]
```diff
@@ -1,13 +1,20 @@
...
+import Button from '@material-ui/core/Button';
+import { RecipesDownloader } from './RecipesDownloader';
+import RecipesPagination from './RecipesPagination';
 
-const useStyles = makeStyles(() =>
+const useStyles = makeStyles((theme: Theme) =>
   createStyles({
+    root: {
+      display: 'flex',
+      '& > * + *': {
```
Comment (Aliath) [inline src/components/RecipesList/RecipesList.tsx:14]: selector specifity? 🧐

File: src/components/RecipesList/RecipesList.tsx [hunk 10]
```diff
@@ -19,36 +26,89 @@ const useStyles = makeStyles(() =>
...
   const getData = useCallback(async () => {
+    setLoading(true);
     const recipesList = await getRecipes();
     setRecipes(recipesList.data);
+    setLoading(false);
   }, []);
 
   useEffect(() => {
     getData();
   }, []);
```
Comment (Aliath) [inline src/components/RecipesList/RecipesList.tsx:43]: `getData` is missing in dependency array

**Review by Aliath — CHANGES_REQUESTED** [review]

File: src/components/RecipesList/RecipesList.tsx (hunk 6 above)
Comment (TomaszLaski) [reply src/components/RecipesList/RecipesList.tsx:106]: "downloading" recipes from api, I couldn't think of another name, I didn't want to waste time thinking about it, so I called it that way.

File: src/components/RecipesList/RecipesList.tsx (hunk 9 above)
Comment (TomaszLaski) [reply src/components/RecipesList/RecipesList.tsx:14]: actually I got this from material ui example : https://codesandbox.io/s/y9rpj?file=/demo.tsx

File: src/components/RecipesList/RecipesList.tsx [hunk 11]
```diff
@@ -19,36 +26,92 @@ const useStyles = makeStyles(() =>
...
+            variant="outlined"
+            color="primary"
+            onClick={() => setRecipesPerPage(allRecipes)}
+          >
+            Show all Recipes
+          </Button>
+        </div>
+        <RecipesListFetcher loading={loading} recipes={currentRecipes} />
+        <Pagination
+          count={Math.ceil(allRecipes / recipesPerPage)}
```
Comment (Aliath) [inline src/components/RecipesList/RecipesList.tsx:108]: Let's define it as const before `return` statement - for example:
```tsx
const totalPages = Math.ceil(allRecipes / recipesPerPage);
// ...
<Pagination
          count={totalPages}
```

File: src/components/RecipesList/RecipesList.tsx (hunk 5 above)
Comment (Aliath) [reply src/components/RecipesList/RecipesList.tsx:77]: 👀

File: src/components/RecipesList/RecipesListFetcher.tsx [hunk 12]
```diff
@@ -0,0 +1,63 @@
+import React from 'react';
+import ListItem from '@material-ui/core/ListItem';
+import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
+import ListItemText from '@material-ui/core/ListItemText';
+import { List, Grid } from '@material-ui/core';
+import Paper from '@material-ui/core/Paper';
+import { Recipe } from 'services/recipes';
+import CircularProgress from '@material-ui/core/CircularProgress';
+
+interface Props {
```
Comment (Aliath) [inline src/components/RecipesList/RecipesListFetcher.tsx:10]: `Props` is not enough specific

File: src/components/RecipesList/RecipesListFetcher.tsx [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    },
+    textList: {
+      textAlign: 'center',
+      width: '100%',
+      padding: '0.3rem',
+    },
+  })
+);
+
+export const RecipesListFetcher: React.FC<Props> = ({ loading, recipes }) => {
```
Comment (Aliath) [inline src/components/RecipesList/RecipesListFetcher.tsx:31]: Actually it does not fetch recipes 😄 I think this piece of code (whole component) should be inside `RecipesList.tsx`

File: src/components/RecipesList/RecipesListFetcher.tsx [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+          <h2>Loading...</h2>
+          <CircularProgress />
+        </Grid>
+      </div>
+    );
+  }
+
+  return (
+    <List>
+      {recipes.map((recipe: any) => {
```
Comment (Aliath) [inline src/components/RecipesList/RecipesListFetcher.tsx:52]: `any` 👀

**Review by Aliath — CHANGES_REQUESTED** [review]

File: src/components/RecipesList/RecipesList.tsx [hunk 15]
```diff
@@ -49,6 +117,15 @@ export const RecipesList: React.FC = () => {
             );
           })}
         </List>
+        <Pagination
+          count={totalPages}
+          color="secondary"
+          variant="outlined"
+          page={1}
+          onChange={(e, value: React.SetStateAction<number>) =>
```
Comment (Aliath) [inline src/components/RecipesList/RecipesList.tsx:125]: There's no need to specify type of `value` - it should be inferred

File: src/components/RecipesList/RecipesList.tsx [hunk 16]
```diff
@@ -19,27 +28,86 @@ const useStyles = makeStyles(() =>
...
   const recipesStyles = useStyles();
   const [recipes, setRecipes] = useState<Recipe[]>([]);
+  const [loading, setLoading] = useState<boolean>(false);
+  const [currentRecipe, setCurrentRecipe] = useState<number>(1);
+  const [recipesPerPage, setRecipesPerPage] = useState<number>(10);
+  const [inputValue, setInputValue] = useState<number>(10);
 
   const getData = useCallback(async () => {
+    setLoading(true);
     const recipesList = await getRecipes();
```
Comment (Aliath) [inline src/components/RecipesList/RecipesList.tsx:38]: What about to wrap `getRecipes` with try/catch blocks? Thanks to it your state can be more explicit.

Let's assume that http request failed - you have to show a feedback about it. So you can create a state that will store state of request: `loading`, `success`, `failure`

Please take a look [on it](https://github.com/knowde-academy/team-1-fe-react/pull/4/files#diff-59d27d293b14d8e5ad67f81192cdc3e750c553a0b1f32165586f333e37468d79R44)

**Review by Aliath — CHANGES_REQUESTED** [review]

File: src/components/RecipesList/RecipesList.tsx [hunk 17]
```diff
@@ -19,27 +28,91 @@ const useStyles = makeStyles(() =>
...
-  }, []);
+  }, [getData]);
+
+  const allRecipes = recipes.length;
+  const indexOfLastRecipe = currentRecipe * recipesPerPage;
+  const indexOfFirstRecipe = indexOfLastRecipe - recipesPerPage;
+  const currentRecipes = recipes.slice(indexOfFirstRecipe, indexOfLastRecipe);
+  const totalPages = Math.ceil(allRecipes / recipesPerPage);
+
+  if (requestState === 'loading' || requestState === 'error') {
```
Comment (Aliath) [inline src/components/RecipesList/RecipesList.tsx:58]: Why does `error` state displays spinner instead of negative feedback? 😄

**Review by Aliath — CHANGES_REQUESTED** [review]

**Review by Aliath — APPROVED** [review]

**Pull request merged**